src/librustdoc/clean/mod.rs

```diff
@@ -1813,55 +1755,57 @@ impl Clean<Visibility> for hir::Visibility<'_> {
             hir::VisibilityKind::Restricted { ref path, .. } => {
                 let path = path.clean(cx);
                 let did = register_res(cx, path.res);
-                Visibility::Restricted(did, path)
+                Visibility::Restricted(did, cx.tcx.def_path(did))
             }
         }
     }
 }
 
 impl Clean<Visibility> for ty::Visibility {
-    fn clean(&self, _: &DocContext<'_>) -> Visibility {
-        if *self == ty::Visibility::Public { Public } else { Inherited }
+    fn clean(&self, cx: &DocContext<'_>) -> Visibility {
+        match *self {
+            ty::Visibility::Public => Visibility::Public,
+            ty::Visibility::Restricted(module) => {
+                if module.is_top_level_module() {
+                    Visibility::Crate
+                } else {
+                    Visibility::Restricted(module, cx.tcx.def_path(module))
+                }
+            }
+            ty::Visibility::Invisible => Visibility::Inherited,
+        }
```

Comment:
You can use the logic from

https://github.com/rust-lang/rust/blob/master/compiler/rustc_privacy/src/lib.rs#L1741-L1753

to prefer "" to "pub(crate)"/"pub(in path)", and "pub(crate)" to "pub(in path)".

Reply:
It can be applied to `impl Clean<Visibility> for hir::Visibility<'_> { ... }` as well since you are using def-paths now anyway.

Reply:
That works on the HIR, so I need a way to get the parent module of a DefId. I have some code for that in https://github.com/rust-lang/rust/blob/98d66340d6e63eda115afc8b0da1d87965881936/src/librustdoc/passes/collect_intra_doc_links.rs#L810-L828, do you think it makes sense to turn that into a query? Or should I just make a free function in rustdoc?

Reply:
There's a readily available method for obtaining a parent module of `HirId` - `tcx.parent_module(hir_id)` (and you don't need this functionality for other crates).

Reply:
> and you don't need this functionality for other crates

Hmm, I think there's a bug in rustc somewhere then: [`impl MetadataExt for Metadata`](https://doc.rust-lang.org/nightly/src/std/os/linux/fs.rs.html#327) has a `Visibility::Restricted` even though it's from another crate.

```
DEBUG rustdoc::clean::inline build_impl: impl Some(DefId(1:7424 ~ std[7f8a]::os::linux::fs::MetadataExt)) for Some(DefId(1:7039 ~ std[7f8a]::fs::Metadata))
DEBUG rustdoc::clean::types name=None, def_id=DefId(1:7442 ~ std[7f8a]::os::linux::fs::{impl#0})
thread 'rustc' panicked at 'only local items can have restricted visibility (while calculating vis of DefId(1:7420 ~ std[7f8a]::os::linux::fs))', src/librustdoc/clean/mod.rs:1762:51
```

This is from local changes so the backtrace doesn't match the PR - let me know if I should push the broken code.

<details><summary>Backtrace</summary>

```
stack backtrace:
   0: rust_begin_unwind
   1: std::panicking::begin_panic_fmt
   2: rustdoc::clean::ty_visibility::{{closure}}
   3: rustdoc::clean::ty_visibility
   4: rustdoc::clean::types::Item::from_def_id_and_parts
   5: rustdoc::clean::inline::build_impl
   6: rustc_session::utils::<impl rustc_session::session::Session>::time
   7: rustdoc::passes::collect_trait_impls::collect_trait_impls
   8: rustc_session::utils::<impl rustc_session::session::Session>::time
   9: rustdoc::core::run_global_ctxt
  10: rustc_interface::passes::QueryContext::enter
  11: rustc_session::utils::<impl rustc_session::session::Session>::time
  12: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::enter
  13: rustc_span::with_source_map
  14: rustc_interface::interface::create_compiler_and_run
  15: rustdoc::core::run_core
  16: rustdoc::main_options
  17: scoped_tls::ScopedKey<T>::set
  18: rustc_span::with_session_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

The `def_id` is coming from https://github.com/rust-lang/rust/blob/master/src/librustdoc/clean/inline.rs#L341.
```

</details>